README.md
Rakam iOS SDK
====================
An iOS SDK for tracking events to [Rakam](http://www.rakam.com).

[![CocoaPods](https://img.shields.io/cocoapods/v/rakam-ios.svg?style=flat)](http://cocoadocs.org/docsets/Rakam-iOS/)

A [demo application](https://github.com/rakam-io/rakam-ios-demo) is available to show a simple integration.

See our [SDK documentation](https://rawgit.com/rakam-io/rakam-ios/v4.0.1/documentation/html/index.html) for a description of all available SDK methods and classes.

Our iOS SDK also supports tvOS. See [below](https://github.com/rakam-io/rakam-ios#tvos) for more information.

# Setup #
1. If you haven't already, go to https://rakam.com and register for an account. You will receive an API Key.

2. [Download the source code](https://github.com/rakam/Rakam-iOS/archive/master.zip) and extract the zip file.

    Alternatively, you can pull directly from GitHub. If you use CocoaPods, add the following line to your Podfile: `pod 'Rakam-iOS', '~> 4.0.1'`. If you are using CocoaPods, you may skip steps 3 and 4.

3. Copy the `Rakam` sub-folder into the source of your project in Xcode. Check "Copy items into destination group's folder (if needed)".

4. Rakam's iOS SDK requires the SQLite library, which is included in iOS but requires an additional build flag to enable. In your project's `Build Settings` and your Target's `Build Settings`, under `Linking` -> `Other Linker Flags`, add the flag `-lsqlite3.0`.

5. In every file that uses analytics, import Rakam.h at the top:
    ``` objective-c
    #import "Rakam.h"
    ```

6. In the application:didFinishLaunchingWithOptions: method of your YourAppNameAppDelegate.m file, initialize the SDK:
    ``` objective-c
    [[Rakam instance] initializeApiKey:[NSURL URLWithString:@"YOUR_RAKAM_API"] : @"YOUR_API_KEY_HERE"];
    ```

7. To track an event anywhere in the app, call:
    ``` objective-c
    [[Rakam instance] logEvent:@"EVENT_IDENTIFIER_HERE"];
    ```

8. Events are saved locally. Uploads are batched to occur every 30 events and every 30 seconds, as well as on app close. After calling logEvent in your app, you will immediately see data appear on the Rakam Website.

# Tracking Events #

It's important to think about what types of events you care about as a developer. You should aim to track between 20 and 200 types of events on your site. Common event types are actions the user initiates (such as pressing a button) and events you want the user to complete (such as filling out a form, completing a level, or making a payment).

A single call to `logEvent` should not have more than 1000 event properties. Likewise a single call to `setUserProperties` should not have more than 1000 user properties. If the 1000 item limit is exceeded then the properties will be dropped and a warning will be logged. We have put in very conservative estimates for the event and property caps which we don’t expect to be exceeded in any practical use case. If you feel that your use case will go above those limits please reach out to e@rakam.com.

# Tracking Sessions #

A session is a period of time that a user has the app in the foreground. Sessions within 5 minutes of each other are merged into a single session. In the iOS SDK, sessions are tracked automatically. When the SDK is initialized, it determines whether the app is launched into the foreground or background and starts a new session if launched in the foreground. A new session is created when the app comes back into the foreground after being out of the foreground for 5 minutes or more. If the app is in the background and an event is logged, then a new session is created if more than 5 minutes has passed since the app entered the background or when the last event was logged (whichever occured last). Otherwise the background event logged will be part of the current session.

You can adjust the time window for which sessions are extended by changing the variable minTimeBetweenSessionsMillis:
``` objective-c
[Rakam instance].minTimeBetweenSessionsMillis = 30 * 60 * 1000; // 30 minutes
[[Rakam instance] initializeApiKey:[NSURL URLWithString:@"YOUR_RAKAM_API_URL"] : @"YOUR_API_KEY_HERE"];
```

By default start and end session events are no longer sent. To renable add this line before initializing the SDK:
``` objective-c
[[Rakam instance] setTrackingSessionEvents:YES];
[[Rakam instance] initializeApiKey:[NSURL URLWithString:@"YOUR_RAKAM_API_URL"] : @"YOUR_API_KEY_HERE"];
```

You can also log events as out of session. Out of session events have a session_id of -1 and are not considered part of the current session, meaning they do not extend the current session. This might be useful for example if you are logging events triggered by push notifications. You can log events as out of session by setting input parameter outOfSession to true when calling logEvent.

``` objective-c
[[Rakam instance] logEvent:@"EVENT_IDENTIFIER_HERE" withEventProperties:nil outOfSession:true];
```

You can also log identify events as out of session by setting input parameter `outOfSession` to `YES` when calling identify:

``` objective-c
RakamIdentify *identify = [[RakamIdentify identify] set:@"key" value:@"value"];
[[Rakam instance] identify:identify outOfSession:YES];
```

### Getting the Session Id ###

You can use the helper method `getSessionId` to get the value of the current sessionId:
``` objective-c
long long sessionId = [[Rakam instance] getSessionId];
```

# Setting Custom User IDs #

If your app has its own login system that you want to track users with, you can call `setUserId:` at any time:

``` objective-c
[[Rakam instance] setUserId:@"USER_ID_HERE"];
```

You can also add the user ID as an argument to the `initializeApiKey:` call:

``` objective-c
[[Rakam instance] initializeApiKey:[NSURL URLWithString:@"YOUR_RAKAM_API_URL"] : @"YOUR_API_KEY_HERE" userId:@"USER_ID_HERE"];
```

### Logging Out and Anonymous Users ###
If a user logs out, or you want to log the events under an anonymous user, you need to do 2 things: 1) set the userId to `nil` 2) regenerate a new deviceId. After doing that, events coming from the current user/device will appear as a brand new user in Rakam dashboards. Note: if you choose to do this, you won't be able to see that the 2 users were using the same device.

``` objective-c
[[Rakam instance] setUserId:nil];  // not string nil
[[Rakam instance] regenerateDeviceId];
```


# Setting Event Properties #

You can attach additional data to any event by passing a NSDictionary object as the second argument to logEvent:withEventProperties:

``` objective-c
NSMutableDictionary *eventProperties = [NSMutableDictionary dictionary];
[eventProperties setValue:@"VALUE_GOES_HERE" forKey:@"KEY_GOES_HERE"];
[[Rakam instance] logEvent:@"Compute Hash" withEventProperties:eventProperties];
```

Note: the keys should be of type NSString, and the values should be of type NSString, NSNumber, NSArray, NSDictionary, or NSNull. You will see a warning if you try to use an unsupported type.

# User Properties and User Property Operations #

The SDK supports the operations set, setOnce, unset, and add on individual user properties. The operations are declared via a provided `RakamIdentify` interface. Multiple operations can be chained together in a single `RakamIdentify` object. The `RakamIdentify` object is then passed to the Rakam client to send to the server. The results of the operations will be visible immediately in the dashboard, and take effect for events logged after. Note, each
operation on the `RakamIdentify` object returns the same instance, allowing you to chain multiple operations together.

To use the `RakamIdentify` interface, you will first need to include the header:
``` objective-c
#import "RakamIdentify.h"
```

1. `set`: this sets the value of a user property.

    ``` objective-c
    RakamIdentify *identify = [[[RakamIdentify identify] set:@"gender" value:@"female"] set:@"age" value:[NSNumber numberForInt:20]];
    [[Rakam instance] identify:identify];
    ```

2. `setOnce`: this sets the value of a user property only once. Subsequent `setOnce` operations on that user property will be ignored. In the following example, `sign_up_date` will be set once to `08/24/2015`, and the following setOnce to `09/14/2015` will be ignored:

    ``` objective-c
    RakamIdentify *identify1 = [[RakamIdentify identify] setOnce:@"sign_up_date" value:@"09/06/2015"];
    [[Rakam instance] identify:identify1];

    RakamIdentify *identify2 = [[RakamIdentify identify] setOnce:@"sign_up_date" value:@"10/06/2015"];
    [[Rakam instance] identify:identify2];
    ```

3. `unset`: this will unset and remove a user property.

    ``` objective-c
    RakamIdentify *identify = [[[RakamIdentify identify] unset:@"gender"] unset:@"age"];
    [[Rakam instance] identify:identify];
    ```


CHANGELOG.md
## Unreleased

### 3.14.0 (February 2, 2017)

* Add support for enabling SSL-pinning via Cocoapods. Thanks to @aaronwasserman for the PR. See [Readme](https://github.com/rakam/rakam-ios#ssl-pinning) for more information.

### 3.13.0 (January 30, 2017)

* Add support for tvOS. Thanks to @gabek for the original PR. See [Readme](https://github.com/rakam/Rakam-iOS#tvos) for more information.
* Bump iOS minimum deployment target to 6.0.
* Update device list. Thanks to @subbotkin for the PR.

### 3.12.1 (December 15, 2016)

* Fix bug where `regenerateDeviceId` was not being run on background thread.
* `[RKMDeviceInfo generateUUID]` should be a static method.

### 3.12.0 (December 5, 2016)

* Add helper method to regenerate a new random deviceId. This can be used in conjunction with `setUserId:nil` to anonymize a user after they log out. Note this is not recommended unless you know what you are doing. See [Readme](https://github.com/rakam/Rakam-iOS#logging-out-and-anonymous-users) for more information.

### 3.11.1 (November 7, 2016)

* Allow `logEvent` with a custom long long timestamp (milliseconds since epoch). See [iOS documentation](https://rawgit.com/rakam/Rakam-iOS/v3.11.1/documentation/html/Classes/Rakam.html#//api/name/logEvent:withEventProperties:withGroups:withLongLongTimestamp:outOfSession:) for more details.

### 3.11.0 (November 7, 2016)

* Allow `logEvent` with a custom timestamp (milliseconds since epoch). If the timestamp value is `nil`, then the event is timestamped with the current time. If setting a custom timestamp, you should use `[NSNumber numberWithLongLong:milliseconds]`. See [iOS documentation](https://rawgit.com/rakam/Rakam-iOS/master/documentation/html/Classes/Rakam.html#//api/name/logEvent:withEventProperties:withGroups:withTimestamp:outOfSession:) for more details.

### 3.10.1 (October 31, 2016)

* Enable "Weak References in Manual Retain Release" to fix build errors in Xcode 7.3 and up.

### 3.10.0 (October 26, 2016)

* Add ability to log identify events outOfSession, this is useful for updating user properties without triggering session-handling logic. See [Readme](https://github.com/rakam/Rakam-iOS#tracking-sessions) for more information.

### 3.9.0 (October 7, 2016)

* Switch to unarchiving unsent events archive file with `[NSKeyedUnarchiver unarchiveObjectWithFile]` to iOS 9's `[NSKeyedUnarchiver unarchiveTopLevelObjectWithData]`. Note: this only affects you if you are *upgrading from an SDK version older than v3.1.0 straight to v3.9.0 or newer*. Users who have not updated to iOS 9.0 or newer will lose any unsent events stored on their devices. This also removes all Objective-C Exceptions (@try/@catch) from the SDK, removing the need to toggle `Enable Objective-C Exceptions` in Xcode.
* Block event property and user property dictionaries that have more than 1000 items. This is to block properties that are set unintentionally (for example in a loop). A single call to `logEvent` should not have more than 1000 event properties. Similarly a single call to `setUserProperties` should not have more than 1000 user properties.

### 3.8.5 (August 29, 2016)

* Fix crash by handling NULL events saved to and fetched from the database.

### 3.8.4 Re-release (August 19, 2016)

* Added support for integration via Carthage. Thanks to @mpurland for the original PR. Thanks to @lexrus for follow up PR to fix framework naming.
* Cleaned up warnings for expression result unused.
* Note if you installed 3.8.4 on August 18, just rerun `pod install` or `carthage update` to pull in the new changes. The re-release was to fix the Carthage framework naming.

### 3.8.3 (July 18, 2016)

* Fix overflow bug for long long values saved to Sqlite DB on 32-bit devices.

### 3.8.2 (July 11, 2016)

* `productId` is no longer a required field for `Revenue` logged via `logRevenue`.
* Fix bug where revenue receipt was being truncated if it was too long (exceeded 1024 characters);

## 3.8.1 (June 14, 2016)

* Allow ability to silence error messages. Note error messages are printed by default. To disable error logging, change `RAKAM_LOG_ERRORS` from `1` to `0` in `Rakam.m`.

## 3.8.0 (June 13, 2016)

* Add support for iOS Extensions. See the [Readme](https://github.com/rakam/rakam-ios#ios-extensions) for instructions, or check out our [iOS-Extension-Demo](https://github.com/rakam/iOS-Extension-Demo). Credit to @andyyc for the original PR.

Rakam/Rakam.h
//
// Rakam.h

#import <Foundation/Foundation.h>
#import "RakamIdentify.h"
#import "RakamRevenue.h"


/**
 Rakam iOS SDK.

 Use the Rakam SDK to track events in your application.

 Setup:

 1. In every file that uses analytics, import Rakam.h at the top `#import "Rakam.h"`
 2. Be sure to initialize the API in your didFinishLaunchingWithOptions delegate `[[Rakam instance] initializeApiKey:[NSURL URLWithString:@"https://app.rakam.io"] : @"YOUR_API_KEY_HERE"];`
 3. Track an event anywhere in the app `[[Rakam instance] logEvent:@"EVENT_IDENTIFIER_HERE"];`
 4. You can attach additional data to any event by passing a NSDictionary object:

        NSMutableDictionary *eventProperties = [NSMutableDictionary dictionary];
        [eventProperties setValue:@"VALUE_GOES_HERE" forKey:@"KEY_GOES_HERE"];
        [[Rakam instance] logEvent:@"Compute Hash" withEventProperties:eventProperties];

 **Note:** you should call SDK methods on an Rakam instance, for example logging events with the default instance: `[[Rakam instance] logEvent:@"testEvent"];`

 **Note:** the SDK supports tracking data to multiple Rakam apps, via separate named instances. For example: `[[Rakam instanceWithName:@"app1"] logEvent:@"testEvent"];` See [Tracking Events to Multiple Apps](https://github.com/rakam-io/rakam-ios#tracking-events-to-multiple-rakam-apps).

 For more details on the setup and usage, be sure to checkout the [README](https://github.com/rakam-io/rakam-ios)
 */
@interface Rakam : NSObject

#pragma mark - Properties

/**-----------------------------------------------------------------------------
* @name Instance Properties
* -----------------------------------------------------------------------------
*/

/**
 API key for your Rakam App.
 */
@property(nonatomic, strong, readonly) NSString *apiKey;
@property(nonatomic, strong, readonly) NSString *apiUrl;

/**
 Identifier for the current user.
 */
@property(nonatomic, strong, readonly) NSString *userId;

/**
 Identifier for the current device.
 */
@property(nonatomic, strong, readonly) NSString *deviceId;

/**
 Name of the SDK instance (ex: no name for default instance, or custom name for a named instance)
 */
@property(nonatomic, strong, readonly) NSString *instanceName;
@property(nonatomic, strong, readonly) NSString *propertyListPath;

/**
 Whether or to opt the current user out of tracking. If true then this blocks the logging of any events and properties, and blocks the sending of events to Rakam servers.
 */
@property(nonatomic, assign) BOOL optOut;


/**-----------------------------------------------------------------------------
 * @name Configurable SDK thresholds and parameters
 * -----------------------------------------------------------------------------
 */

/**
 The maximum number of events that can be stored locally before forcing an upload. The default is 30 events.
 */
@property(nonatomic, assign) int eventUploadThreshold;

/**
 The maximum number of events that can be uploaded in a single request. The default is 100 events.
 */
@property(nonatomic, assign) int eventUploadMaxBatchSize;

/**
 The maximum number of events that can be stored lcoally. The default is 1000 events.
 */
@property(nonatomic, assign) int eventMaxCount;

/**
 The amount of time after an event is logged that events will be batched before being uploaded to the server. The default is 30 seconds.
 */
@property(nonatomic, assign) int eventUploadPeriodSeconds;

/**
 When a user closes and reopens the app within minTimeBetweenSessionsMillis milliseconds, the reopen is considered part of the same session and the session continues. Otherwise, a new session is created. The default is 15 minutes.
 */
@property(nonatomic, assign) long minTimeBetweenSessionsMillis;

/**
 Whether to automatically log start and end session events corresponding to the start and end of a user's session.
 */
@property(nonatomic, assign) BOOL trackingSessionEvents;


#pragma mark - Methods

/**-----------------------------------------------------------------------------
 * @name Fetching Rakam SDK instance
 * -----------------------------------------------------------------------------
 */

/**
 This fetches the default SDK instance. Recommended if you are only logging events to a single app.

 @returns the default Rakam SDK instance
 */
+ (Rakam *)instance;

/**
 This fetches a named SDK instance. Use this if logging events to multiple Rakam apps.

 @param instanceName the name of the SDK instance to fetch.

 @returns the Rakam SDK instance corresponding to `instanceName`

 @see [Tracking Events to Multiple Rakam Apps](https://github.com/rakam-io/rakam-ios#tracking-events-to-multiple-rakam-apps)
 */
+ (Rakam *)instanceWithName:(NSString *)instanceName;

/**-----------------------------------------------------------------------------
 * @name Initialize the Rakam SDK with your Rakam API Key
 * -----------------------------------------------------------------------------
 */

/**
 Initializes the Rakam instance with your Rakam API key

 We recommend you first initialize your class within your "didFinishLaunchingWithOptions" method inside your app delegate.

 **Note:** this is required before you can log any events.

 @param apiKey Your Rakam key obtained from your dashboard at https://rakam.com/settings
 */
- (void)initializeApiKey:(NSURL *)apiUrl :(NSString *)apiKey;

/**
 Initializes the Rakam instance with your Rakam API key and sets a user identifier for the current user.

 We recommend you first initialize your class within your "didFinishLaunchingWithOptions" method inside your app delegate.

 **Note:** this is required before you can log any events.

 @param apiKey Your Rakam key obtained from your dashboard at https://rakam.com/settings

 @param userId If your app has its own login system that you want to track users with, you can set the userId.

*/
- (void)initializeApiKey:(NSURL *)apiUrl :(NSString *)apiKey userId:(NSString *)userId;


/**-----------------------------------------------------------------------------
 * @name Logging Events
 * -----------------------------------------------------------------------------
 */

/**
 Tracks an event. Events are saved locally.

 Uploads are batched to occur every 30 events or every 30 seconds (whichever comes first), as well as on app close.

 @param eventType                The name of the event you wish to track.

 @see [Tracking Events](https://github.com/rakam-io/rakam-ios#tracking-events)
 */
- (void)logEvent:(NSString *)eventType;

/**
 Tracks an event. Events are saved locally.

 Uploads are batched to occur every 30 events or every 30 seconds (whichever comes first), as well as on app close.

 @param eventType                The name of the event you wish to track.
 @param eventProperties          You can attach additional data to any event by passing a NSDictionary object with property: value pairs.

 @see [Tracking Events](https://github.com/rakam-io/rakam-ios#tracking-events)
 */
- (void)logEvent:(NSString *)eventType withEventProperties:(NSDictionary *)eventProperties;

/**
 Tracks an event. Events are saved locally.

 Uploads are batched to occur every 30 events or every 30 seconds (whichever comes first), as well as on app close.

 @param eventType                The name of the event you wish to track.
 @param eventProperties          You can attach additional data to any event by passing a NSDictionary object with property: value pairs.
 @param outOfSession             If YES, will track the event as out of session. Useful for push notification events.

 @see [Tracking Events](https://github.com/rakam-io/rakam-ios#tracking-events)
 @see [Tracking Sessions](https://github.com/rakam-io/rakam-ios#tracking-sessions)
 */
- (void)logEvent:(NSString *)eventType withEventProperties:(NSDictionary *)eventProperties outOfSession:(BOOL)outOfSession;

/**
 Tracks an event. Events are saved locally.

 Uploads are batched to occur every 30 events or every 30 seconds (whichever comes first), as well as on app close.

 @param eventType                The name of the event you wish to track.
 @param eventProperties          You can attach additional data to any event by passing a NSDictionary object with property: value pairs.
 @param groups                   You can specify event-level groups for this user by passing a NSDictionary object with groupType: groupName pairs. Note the keys need to be strings, and the values can either be strings or an array of strings.

 @see [Tracking Events](https://github.com/rakam-io/rakam-ios#tracking-events)

 @see [Setting Groups](https://github.com/rakam-io/rakam-ios#setting-groups)
 */
- (void)logEvent:(NSString *)eventType withEventProperties:(NSDictionary *)eventProperties withGroups:(NSDictionary *)groups;

/**
 Tracks an event. Events are saved locally.

 Uploads are batched to occur every 30 events or every 30 seconds (whichever comes first), as well as on app close.

 @param eventType                The name of the event you wish to track.
 @param eventProperties          You can attach additional data to any event by passing a NSDictionary object with property: value pairs.
 @param groups                   You can specify event-level groups for this user by passing a NSDictionary object with groupType: groupName pairs. Note the keys need to be strings, and the values can either be strings or an array of strings.
 @param outOfSession             If YES, will track the event as out of session. Useful for push notification events.

 @see [Tracking Events](https://github.com/rakam-io/rakam-ios#tracking-events)

 @see [Setting Groups](https://github.com/rakam-io/rakam-ios#setting-groups)

 @see [Tracking Sessions](https://github.com/rakam-io/rakam-ios#tracking-sessions)
 */
- (void)logEvent:(NSString *)eventType withEventProperties:(NSDictionary *)eventProperties withGroups:(NSDictionary *)groups outOfSession:(BOOL)outOfSession;

/**
 Tracks an event. Events are saved locally.

 Uploads are batched to occur every 30 events or every 30 seconds (whichever comes first), as well as on app close.

 @param eventType                The name of the event you wish to track.
 @param eventProperties          You can attach additional data to any event by passing a NSDictionary object with property: value pairs.
 @param groups                   You can specify event-level groups for this user by passing a NSDictionary object with groupType: groupName pairs. Note the keys need to be strings, and the values can either be strings or an array of strings.
 @param longLongTimestamp        You can specify a custom timestamp by passing the milliseconds since epoch UTC time as a long long.
 @param outOfSession             If YES, will track the event as out of session. Useful for push notification events.

 @see [Tracking Events](https://github.com/rakam-io/rakam-ios#tracking-events)

 @see [Setting Groups](https://github.com/rakam-io/rakam-ios#setting-groups)

 @see [Tracking Sessions](https://github.com/rakam-io/rakam-ios#tracking-sessions)
 */
- (void)logEvent:(NSString *)eventType withEventProperties:(NSDictionary *)eventProperties withGroups:(NSDictionary *)groups withLongLongTimestamp:(long long)longLongTimestamp outOfSession:(BOOL)outOfSession;

/**
 Tracks an event. Events are saved locally.

 Uploads are batched to occur every 30 events or every 30 seconds (whichever comes first), as well as on app close.

 @param eventType                The name of the event you wish to track.
 @param eventProperties          You can attach additional data to any event by passing a NSDictionary object with property: value pairs.
 @param groups                   You can specify event-level groups for this user by passing a NSDictionary object with groupType: groupName pairs. Note the keys need to be strings, and the values can either be strings or an array of strings.
 @param timestamp                You can specify a custom timestamp by passing an NSNumber representing the milliseconds since epoch UTC time. We recommend using [NSNumber numberWithLongLong:milliseconds] to create the value. If nil is passed in, then the event will be timestamped with the current time.
 @param outOfSession             If YES, will track the event as out of session. Useful for push notification events.

 @see [Tracking Events](https://github.com/rakam-io/rakam-ios#tracking-events)

 @see [Setting Groups](https://github.com/rakam-io/rakam-ios#setting-groups)

 @see [Tracking Sessions](https://github.com/rakam-io/rakam-ios#tracking-sessions)
 */
- (void)logEvent:(NSString *)eventType withEventProperties:(NSDictionary *)eventProperties withGroups:(NSDictionary *)groups withTimestamp:(NSNumber *)timestamp outOfSession:(BOOL)outOfSession;

/**-----------------------------------------------------------------------------
 * @name Logging Revenue
 * -----------------------------------------------------------------------------
 */

/**
 Tracks revenue - API v2. This uses the `RKMRevenue` object to store transaction properties such as quantity, price, and revenue type. This is the recommended method for tracking revenue in Rakam.

 For validating revenue, make sure the receipt data is set on the RKMRevenue object.

 To track revenue from a user, create an RKMRevenue object each time the user generates revenue, and set the revenue properties (productIdentifier, price, quantity). logRevenue: takes in an RKMRevenue object. This allows us to automatically display data relevant to revenue on the Rakam website, including average revenue per daily active user (ARPDAU), 7, 30, and 90 day revenue, lifetime value (LTV) estimates, and revenue by advertising campaign cohort and daily/weekly/monthly cohorts.

 @param revenue RKMRevenue object       revenue object contains all revenue information

 @see [Tracking Revenue](https://github.com/rakam-io/rakam-ios#tracking-revenue)
 */
- (void)logRevenue:(RakamRevenue *)revenue;

/**-----------------------------------------------------------------------------
 * @name User Properties and User Property Operations
 * -----------------------------------------------------------------------------
 */

/**
 Update user properties using operations provided via Identify API.

 To update user properties, first create an RKMIdentify object. For example if you wanted to set a user's gender, and then increment their karma count by 1, you would do:

    RKMIdentify *identify = [[[RKMIdentify identify] set:@"gender" value:@"male"] add:@"karma" value:[NSNumber numberWithInt:1]];

 Then you would pass this RKMIdentify object to the identify function to send to the server:

    [[Rakam instance] identify:identify];

 @param identify                   An RKMIdentify object with the intended user property operations

 @see [User Properties and User Property Operations](https://github.com/rakam-io/rakam-ios#user-properties-and-user-property-operations)

 */

- (void)identify:(RakamIdentify *)identify;

/**
 Update user properties using operations provided via Identify API. If outOfSession is `YES` then the identify event is logged with a session id of -1 and does not trigger any session-handling logic.

 To update user properties, first create an RKMIdentify object. For example if you wanted to set a user's gender, and then increment their karma count by 1, you would do:

 RKMIdentify *identify = [[[RKMIdentify identify] set:@"gender" value:@"male"] add:@"karma" value:[NSNumber numberWithInt:1]];

 Then you would pass this RKMIdentify object to the identify function to send to the server:

 [[Rakam instance] identify:identify outOfSession:YES];

 @param identify                   An RKMIdentify object with the intended user property operations
 @param outOfSession               Whether to log identify event out of session

 @see [User Properties and User Property Operations](https://github.com/rakam-io/rakam-ios#user-properties-and-user-property-operations)

 */

- (void)identify:(RakamIdentify *)identify outOfSession:(BOOL)outOfSession;

/**

 Adds properties that are tracked on the user level.

 **Note:** Property keys must be <code>NSString</code> objects and values must be serializable.

 @param userProperties          An NSDictionary containing any additional data to be tracked.

 @see [Setting Multiple Properties with setUserProperties](https://github.com/rakam-io/rakam-ios#setting-multiple-properties-with-setuserproperties)
 */
- (void)setUserProperties:(NSDictionary *)userProperties;

/**
 Clears all properties that are tracked on the user level.

 **Note: the result is irreversible!**

 @see [Clearing user properties](https://github.com/rakam-io/rakam-ios#clearing-user-properties-with-clearuserproperties)
 */

- (void)clearUserProperties;

/**-----------------------------------------------------------------------------
 * @name Setting User and Device Identifiers
 * -----------------------------------------------------------------------------
 */

/**
 Sets the userId.

 @param userId                  If your app has its own login system that you want to track users with, you can set the userId.

 @see [Setting Custom UserIds](https://github.com/rakam-io/rakam-ios#setting-custom-user-ids)
 */
- (void)setUserId:(NSString *)userId;

/**
 Sets the deviceId.

 **NOTE: not recommended unless you know what you are doing**

 @param deviceId                  If your app has its own system for tracking devices, you can set the deviceId.

 @see [Setting Custom Device Ids](https://github.com/rakam-io/rakam-ios#custom-device-ids)
 */
- (void)setDeviceId:(NSString *)deviceId;

/**-----------------------------------------------------------------------------
 * @name Configuring the SDK instance
 * -----------------------------------------------------------------------------
 */

/**
 Enables tracking opt out.

 If the user wants to opt out of all tracking, use this method to enable opt out for them. Once opt out is enabled, no events will be saved locally or sent to the server. Calling this method again with enabled set to NO will turn tracking back on for the user.

 @param enabled                  Whether tracking opt out should be enabled or disabled.
 */
- (void)setOptOut:(BOOL)enabled;

/**
 Disables sending logged events to Rakam servers.

 If you want to stop logged events from being sent to Rakam severs, use this method to set the client to offline. Once offline is enabled, logged events will not be sent to the server until offline is disabled. Calling this method again with offline set to NO will allow events to be sent to server and the client will attempt to send events that have been queued while offline.

 @param offline                  Whether logged events should be sent to Rakam servers.
 */
- (void)setOffline:(BOOL)offline;

/**
 Enables location tracking.

 If the user has granted your app location permissions, the SDK will also grab the location of the user. Rakam will never prompt the user for location permissions itself, this must be done by your app.

 **Note:** the user's location is only fetched once per session. Use `updateLocation` to force the SDK to fetch the user's latest location.
 */
- (void)enableLocationListening;

/**
 Disables location tracking. If you want location tracking disabled on startup of the app, call disableLocationListening before you call initializeApiKey.
 */
- (void)disableLocationListening;

/**
 Forces the SDK to update with the user's last known location if possible.

 If you want to manually force the SDK to update with the user's last known location, call updateLocation.
 */
- (void)updateLocation;

/**
 Uses advertisingIdentifier instead of identifierForVendor as the device ID

 Apple prohibits the use of advertisingIdentifier if your app does not have advertising. Useful for tying together data from advertising campaigns to anlaytics data.

 **NOTE:** Must be called before initializeApiKey: is called to function.
 */
- (void)useAdvertisingIdForDeviceId;

/**-----------------------------------------------------------------------------
 * @name Other Methods
 * -----------------------------------------------------------------------------
 */

/**
 Prints the number of events in the queue.

 Debugging method to find out how many events are being stored locally on the device.
 */
- (void)printEventsCount;

/**
 Fetches the deviceId, a unique identifier shared between multiple users using the same app on the same device.

 @returns the deviceId.
 */
- (NSString *)getDeviceId;

/**
 Regenerates a new random deviceId for current user. Note: this is not recommended unless you know what you are doing. This can be used in conjunction with setUserId:nil to anonymize users after they log out. With a nil userId and a completely new deviceId, the current user would appear as a brand new user in dashboard.

 @see [Logging Out Users](https://github.com/rakam-io/rakam-ios#logging-out-and-anonymous-users)
 */
- (void)regenerateDeviceId;

/**
 Fetches the current sessionId, an identifier used by Rakam to group together events tracked during the same session.

 @returns the current session id

 @see [Tracking Sessions](https://github.com/rakam-io/rakam-ios#tracking-sessions)
 */
- (long long)getSessionId;

/**
 Manually forces the instance to immediately upload all unsent events.

 Events are saved locally. Uploads are batched to occur every 30 events and every 30 seconds, as well as on app close. Use this method to force the class to immediately upload all queued events.
 */
- (void)uploadEvents;


#pragma mark - Deprecated methods

- (void)initializeApiKey:(NSURL *)apiUrl :(NSString *)apiKey userId:(NSString *)userId startSession:(BOOL)startSession __attribute((deprecated()));

- (void)startSession __attribute((deprecated()));

+ (void)initializeApiKey:(NSURL *)apiUrl :(NSString *)apiKey __attribute((deprecated()));

+ (void)initializeApiKey:(NSURL *)apiUrl :(NSString *)apiKey userId:(NSString *)userId __attribute((deprecated()));

+ (void)logEvent:(NSString *)eventType __attribute((deprecated()));

+ (void)logEvent:(NSString *)eventType withEventProperties:(NSDictionary *)eventProperties __attribute((deprecated()));

+ (void)uploadEvents __attribute((deprecated()));

+ (void)setUserProperties:(NSDictionary *)userProperties __attribute((deprecated()));

+ (void)setUserId:(NSString *)userId __attribute((deprecated()));

+ (void)enableLocationListening __attribute((deprecated()));

+ (void)disableLocationListening __attribute((deprecated()));

+ (void)useAdvertisingIdForDeviceId __attribute((deprecated()));

+ (void)printEventsCount __attribute((deprecated()));

+ (NSString *)getDeviceId __attribute((deprecated()));
@end

#pragma mark - constants

extern NSString *const kRKMSessionStartEvent;
extern NSString *const kRKMSessionEndEvent;
extern NSString *const kRKMRevenueEvent;
